gputop/remote/gputop-nodejs-csv.js

```diff
@@ -20,21 +24,124 @@ GPUTopNodeJSUI.prototype.weblog = function(message)
     console.log(message);
 }
 
-GPUTopNodeJSUI.prototype.display_features = function(message)
+GPUTopNodeJSUI.prototype.update_features = function(features)
 {
+    console.log(features);
+
+    features.supported_oa_query_guids.forEach(function (guid, i, a) {
+        var metric = gputop.get_map_metric(guid);
+        metric.counters_.forEach(function (i, j, k) {
+            metric.counters_[j].record_data = true;
+        });
+    });
+
+    gputop.open_oa_metric_set({guid: features.supported_oa_query_guids[0]});
+    supported_oa_query_guids = features.supported_oa_query_guids;
+}
+
+GPUTopNodeJSUI.prototype.queue_redraw = function()
+{
+    if (supported_oa_query_guids !== undefined)
+        var metric = gputop.get_map_metric(supported_oa_query_guids[0]);
+}
+
+GPUTopNodeJSUI.prototype.load_metrics_panel = function(open_query)
+{
+    open_query();
+}
+
+GPUTopNodeJSUI.prototype.update_slider_period = function(period)
+{
+}
+
+GPUTopNodeJSUI.prototype.render_bars = function()
+{
+}
+
+GPUTopNodeJSUI.prototype.log = function(level, message)
+{
+    console.log(level);
     console.log(message);
 }
 
 var gputop_ui = new GPUTopNodeJSUI();
 
-function gputop_ready(gputop)
+function gputop_ready(gputop_inst)
 {
+    gputop = gputop_inst;
     gputop.connect();
 }
 
-http.get("http://localhost:7890/gputop.js", function(response) {
+function gputop_is_demo () {
+    return false;
+}
+
+function on_gputop_web_ready(gputop_web) {
+    http.get("http://localhost:7890/gputop.js", function(response) {
+        var gputopjs = "";
+        response.setEncoding('utf8');
+        response.on('data', function(data) {
+            gputopjs += data;
+        });
+        response.on('end', function () {
+            eval(gputop_web);
+            eval(gputopjs);
+        });
+    });
+}
+
+http.get("http://localhost:7890/gputop-web.js", function(response) {
+    var gputop_web = "";
     response.setEncoding('utf8');
     response.on('data', function(data) {
-        eval(data);
+        gputop_web += data;
+    });
+    response.on('end', function () {
+        on_gputop_web_ready(gputop_web);
     });
 });
+
+function gputop_clean_exit()
+{
+    if (gputop !== undefined && supported_oa_query_guids !== undefined) {
+        var metric = gputop.get_map_metric(supported_oa_query_guids[0]);
+        var fs = require('fs');
```

Comment:
I think require is meant to be considered like #include or import etc and usually be done at the top of the file, outside of any function. googling to see if this is the best practice, I came across this: http://justbuildsomething.com/node-js-best-practices/#2 which justifies putting at the top for the sake of catching any require() errors up front (though not necessarily as major issue in our case, compared to their webserver example.)